### DISCUSSION TITLE
How to keep the arity on the first line for every Clojure fn?

### DISCUSSION THREAD

**Discussion by zikajk** [joaotavora/eglot#1181] [Q&A] [answered]
**EDIT:The original question was 'How to get eldoc-documentation-strategy right' but it was my wrong assumption about the problem....**

Hi, I am using Eglot and Emacs@28 (MacPort). 

1. When i move cursor on some code, I often get combination of eglot-hover-eldoc-function + eglot-signature-eldoc-function in eldoc buffer. 

2. I suppose it is because Eglot changes eldoc-documentation-strategy to eldoc-documentation-compose. 

3. What can I do about it? I tried a lot of things, but I  am not able to get only one documentation for every symbol.

![image](https://user-images.githubusercontent.com/40757859/224690661-18120a58-0226-44db-aa2e-27a9e50f480f.png)

**Comment by joaotavora** [comment]
You can use `eglot-managed-mode-hook` to do adjustments to `eldoc-documentation-functions`, perhaps removing `eldoc-hover-eldoc-function`, but then you'll lose it in `M-x eldoc` output as well.

Now to your questions

1.  Yes, that's what's supposed to work.,  But as I wrote above you can adjust it.
2. Yes, but this makes sense for Eglot generally.  But you can _also_ tweak that.
3. I don't know what you mean.  It would seem you want the echo area to not popup hysterically with long hover information.  You can use `eglot-echo-area-use-multiline-p` and set it to 1 or 2.
4. 
This is a pain point that I am working on for the next Eglot version.  The only strategy that really makes sense is `eldoc-documentation-compose` but `eldoc.el` has learn to evolve to not allow enormous documentation to make it into the echo area outlet, (which is only one of its outlets -- see `eldoc-display-functions`), but is not suitable for very long multi-line strings.

**Reply by zikajk** [reply]
What i meant is to use only eglot-hover-eldoc-function or eglot-signature-eldoc-function but always use one of them 
So if I update the picture, i would like to have hidden the first part as this
![image](https://user-images.githubusercontent.com/40757859/224771704-0a5d21c4-b185-46d0-bbbf-ce86c8868e47.png)

Because with some function, there is a lot of extra (duplicated) info.

**Reply by zikajk** [reply]
If I change `eglot-echo-are-use-multiline-p` then I don't get any information about input parameters...
![image](https://user-images.githubusercontent.com/40757859/224773622-dd784c2b-95a2-4f3e-9d6c-f404c454ee8f.png)

I have a strong feeling that I didnt have this issue when i was using some old eglot version..

**Reply by joaotavora** [reply]
> eglot-hover-eldoc-function or eglot-signature-eldoc-function but always use one of them

What if the two are available? Which one do you want to give priority to?

You must understand, I don't have this language server, and you haven't inequivocally clarified what function returns what.

Additionally, there seems to be some strange interference from your configuration.  Are you using "doom" or something of the sort?

So in this example, what does eglot-hover-eldoc-function produce?  And what does eglot-signature-eldoc-function produce?

> If I change eglot-echo-are-use-multiline-p then I don't get any information about input parameters...

If you change it to what?

**Reply by zikajk** [reply]
When I remove` eldoc-hover-eldoc-function`then I don't get eldoc for some symbols that are missing` eglot-signature-eldoc-function`. So, i would remove it if I can have` eglot-signature-eldoc-function` for everything.
This is the case of `defn` in previous image - this is why I don't get input parameters in one line, `defn` is missing eglot-signature-eldoc-function..

**Reply by joaotavora** [reply]
Then _don't_ remove `eldoc-hover-eldoc-function`,  set `eldoc-echo-area-use-multiple-p` to 1 and perhaps `eldoc-echo-area-display-truncation-message` to `nil`.

**Comment by zikajk** [comment]
1) if two are available, I would like to give priority to `eglot-signature-eldoc-function.`
2) Sure, I'll be happy to add any details I can... The first part in that picture above is `eglot-signature-eldoc-function` and the second one is` eldoc-hover-eldoc-function'. 
3) No, I am using vanilla Emacs
4) If i change it to nil, 1 or 2. 
-----------------------------------
When I do this, then there is still the same issue... 
This is ok:
![image](https://user-images.githubusercontent.com/40757859/224777720-c7760ec2-6961-4a14-ac1a-46809828ab70.png)

This is not:
![image](https://user-images.githubusercontent.com/40757859/224777852-ee9ff495-c295-47c7-8869-8ba37a36f9a8.png)

**Reply by zikajk** [reply]
If I change eldoc-echo-area-use-multiple-p to 3 to see params from this:
![image](https://user-images.githubusercontent.com/40757859/224780519-248fe2fc-4b34-497b-a622-cb630cde69d3.png)
-> 
![image](https://user-images.githubusercontent.com/40757859/224780620-efade9a5-76ad-4b86-bc7a-0eb3681eb000.png)

Then it looks ugly with functions where I get` eglot-signature-eldoc-function` returns something...
![image](https://user-images.githubusercontent.com/40757859/224780862-970dc257-6501-4731-a430-58efa6d1983e.png)

**Reply by joaotavora** [reply]
1.  This already happens
2. I'd prefer if you state _exactly_ what each function produces, to remove any ambiguity. instead of "second one" and "first part". I can guess from the context, but sometimes I guess wrong.
3. That's not what vanilla Emacs looks like.  Vanilla Emacs is what you get by running `emacs -Q`.
4. Use `emacs -Q -f package-initialize`, load up your project, show your configurations of these 2/3 variables, and type `M-x eglot`.  Else it's impossible to know what configuration you're using.

> This is not:

What would you like to see there? Your Clojure LS is informing you about the `defn` definition making symbol because _that's where the current point is_.  What/whose signature would you like to see?  `defn`'s? Or `intern-log-fns`'s? Or something else?  Eglot can't do anything here, you have to take it up with the Clojure LS folks.

**Reply by joaotavora** [reply]
Maybe you want to set `eldoc-documentations-strategy` to `eldoc-documentation-enthusiast`.  I dunno.  I would also ask the Clojure LS people to issue signature help for things like `defn`.

**Reply by zikajk** [reply]
3,4. Will try that - I did, no difference. But as you said, it seems to be related to LS implementation.
5. I would like to see this.. I just switched to lsp-mode to give you an working example. Using the same Clojure LS.
![image](https://user-images.githubusercontent.com/40757859/224786978-ca989a12-21a4-4af1-8c42-20e2feb2b1e5.png)

**Reply by zikajk** [reply]
I think I found it. 
Clojure-LSP does not return :signatures for DEFN (and some other functions). 
But it always return form like this (when eglot-hover-eldoc-function)-> 
So I believe that LSP-Mode uses the second sexp as a signatures... 
I will ask Clojure-LSP guys about this... 

```
[(:language "clojure" :value "clojure.core/defn")
	    (:language "clojure" :value "[name doc-string? attr-map? [params*] prepost-map? body]\n[name doc-string? attr-map? ([params*] prepost-map? body) + attr-map?]")
	    "Same as (def name (fn [params* ] exprs*)) or (def\n    name (fn ([params* ] exprs*)+)) with any doc-string or attrs added\n    to the var metadata. prepost-map defines a map with optional keys\n    :pre and :post that contain collections of pre or post conditions." "/Users/akiz/.m2/repository/org/clojure/clojure/1.11.1/clojure-1.11.1.jar:clojure/core.clj"]

```

**Reply by joaotavora** [reply]
> So I believe that LSP-Mode uses the second sexp as a signatures...

Rather, I think it is not inserting `\n` between the first two, is removing the `\n` in the second, and is inserting `\n` after the second as well (or killing the third entirely). ... lsp-mode has a lot of language specific code and Eglot has none. IMO that's what LSP should be about -- language-agnosticism of clients and editor-agnosticism of servers.  But it means that I do something here it might break other servers, which is typical.  

I'll read the spec closely to see it it says anything about this, but I think the correct thing to do is to ask Clojure-LSP people to issue a signature for `defn`, because _that's_ what you're after.  Anything else is a hack.

**Reply by zikajk** [reply]
I do agree. I have asked about this on Clojure-LSP Slack and I am going to keep it updated here. 
For now, I am fine with default documentation-strategy, `(setq eglot-echo-area-use-multiline-p nil)` and with using eldoc-buffer when I need more info.

**Reply by zikajk** [reply]
If somebody has the same issue.. This is the current "workaround". 
```
(with-eval-after-load 'eglot
       (add-to-list 'eglot-server-programs
                    '((clojure-mode clojurescript-mode clojurec-mode)
                      "clojure-lsp" :initializationOptions (:show-docs-arity-on-same-line? t))))
```

**Reply by joaotavora** [reply]
Good workaround.  It does mean that, for some reason, `clojure-lsp` is returning signature info for some symbols as hover information.  I wonder why they simply don't return that information which they already have in the signature.  

If you have a link to the clojure-lsp conversation on github or somewhere else, please show it.

**Reply by zikajk** [reply]
Here it is. Obviously I'm having trouble explaining the problem :)
https://clojurians.slack.com/archives/CPABC1H61/p1678733411503889

**Comment by joaotavora** [comment, marked as answer]
This comment contains a good workaround: https://github.com/joaotavora/eglot/discussions/1181#discussioncomment-5312502

**Reply by zikajk** [reply]
Another solution is to use in`.lsp/config.edn`
`{:hover {:arity-on-same-line? t}}
`

**Reply by joaotavora** [reply]
Neither of these are actually solutions.  

I also think that soon, unfortunately, they will begin to fail.  I'm thinking of making very long hover information stay out of the echo area.  I'll make it customizable, so you can get back this behaviour,  but my current thinking is that multi-line (say > 2) pieces of documentation do not belong in the Emacs echo area, they go to other eldoc outlets.

**Reply by zikajk** [reply]
I am using this with` (setq eglot-echo-area-use-multiline-p nil)`, so I get only 1 line.
